src/clusterfuzz/_internal/system/archive.py

```diff
@@ -34,6 +33,197 @@
+  """_is_attempting_path_traversal"""
+  absolute_file_path = os.path.join(output_dir, os.path.normpath(filename))
+  real_file_path = os.path.realpath(absolute_file_path)
+
+  if real_file_path == output_dir:
+    # Workaround for https://bugs.python.org/issue28488.
+    # Ignore directories named '.'.
+    return False
+
+  if real_file_path != absolute_file_path:
+    logs.log_error('Directory traversal attempted while unpacking archive %s '
+                   '(file path=%s, actual file path=%s). Aborting.' %
+                   (archive_name, absolute_file_path, real_file_path))
+    return True
+  return False
+
+
+class ArchiveMemberInfo:
+  """ArchiveMemberInfo"""
+
+  def __init__(self, filename, is_dir, file_size, compress_size):
+    self._filename = filename
+    self._is_dir = is_dir
+    self._file_size = file_size
+    self._compress_size = compress_size
+
+  @property
+  def filename(self):
+    return self._filename
+
+  def is_dir(self):
+    raise self._is_dir
+
+  @property
+  def file_size(self):
+    return self._file_size
+
+  @property
+  def compress_size(self):
+    return self._compress_size
+
+
+class ArchiveReader:
+  """ArchiveReader"""
+
+  # This should have a few of the handful functions a tool like unzip has
+  def list_files(self) -> [ArchiveMemberInfo]:
+    raise NotImplementedError
+
+  def extract(self, member, path=None, trusted=False):
+    raise NotImplementedError
+
+  def open(self, member):
+    raise NotImplementedError
+
+  def extractall(self, path=None, members=None, trusted=False) -> None:
+    raise NotImplementedError
+
+  def extracted_size(self, file_match_callback=None):
+    return sum(f.file_size
+               for f in self.list_files()
+               if not file_match_callback or file_match_callback(f.filename))
+
+
+class TarArchiveReader(ArchiveReader):
+  """TarArchiveReader"""
+
+  def __init__(self, archive_path, file_obj=None) -> None:
+    # we need to know whether it's a lzma file or not.
+    archive_type = get_archive_type(archive_path)
+    if file_obj:
+      self.archive = tarfile.open(fileobj=file_obj)
+    else:
+      mode = 'r' if archive_type == ArchiveType.TAR else 'r:xz'
+      self.archive = tarfile.open(archive_path, mode=mode)
+    self.archive_path = archive_path
+
+  def list_files(self) -> [ArchiveMemberInfo]:
+    return [
+        ArchiveMemberInfo(
+            filename=f.name,
+            is_dir=f.isdir(),
+            file_size=f.size,
+            compress_size=f.size) for f in self.archive.getmembers()
+    ]
+
+  def open(self, member):
+    return self.archive.extractfile(member)
+
+  def extract(self, member, path=None, trusted=False):
+    # If the output directory is a symlink, get its actual path since we will be
+    # doing directory traversal checks later when unpacking the archive.
+    output_directory = os.path.realpath(path)
+
+    if not trusted:
+      if (_is_attempting_path_traversal(self.archive_path, output_directory,
```

Comment:
This is not your fault, but do we really want to allow path traversal in *any* case? Why gate it on `trusted`?

Reply:
Agreed. For now, I've let the `trusted` part in. Once this gets merged, I'll change all those annoying things in a second time. The idea is to have sort of a "non functional change" for this CL.

Reply:
+1, let's just remove this concept completely and disallow path traversal in all cases.

Reply:
Totally agree on this. I just would like to do it separately, because zip also does things with symlinks when the archive is trusted, and I am unsure what exactly uses this. I am pretty confident with the current state of this PR, so the idea is to first land this "no functional change" PR, and then land much smaller ones that will modify the behavior slightly to clean things up.